codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/EventStreamMarshallerGenerator.kt

```diff
@@ -99,21 +106,59 @@ class EventStreamMarshallerGenerator(
         }
     }
 
-    private fun RustWriter.serializeEvent(struct: StructureShape) {
-        for (member in struct.members()) {
+    private fun RustWriter.renderMarshallEvent(unionMember: MemberShape, struct: StructureShape) {
+        val headerMembers = struct.members().filter { it.hasTrait<EventHeaderTrait>() }
+        val payloadMember = struct.members().firstOrNull { it.hasTrait<EventPayloadTrait>() }
+        for (member in headerMembers) {
             val memberName = symbolProvider.toMemberName(member)
             val target = model.expectShape(member.target)
-            if (member.hasTrait<EventPayloadTrait>()) {
-                serializeUnionMember(memberName, member, target)
-            } else if (member.hasTrait<EventHeaderTrait>()) {
-                TODO("TODO(EventStream): Implement @eventHeader trait")
-            } else {
-                throw IllegalStateException("Event Stream members must be a header or payload")
+            renderMarshallEventHeader(memberName, member, target)
+        }
+        if (payloadMember != null) {
+            val memberName = symbolProvider.toMemberName(payloadMember)
+            val target = model.expectShape(payloadMember.target)
+            renderMarshallEventPayload("inner.$memberName", payloadMember, target)
+        } else if (headerMembers.isEmpty()) {
+            renderMarshallEventPayload("inner", unionMember, struct)
+        } else {
+            addStringHeader(":content-type", "application/octet-stream".dq() + ".into()")
+            rust("Vec::new()")
+        }
+    }
+
+    private fun RustWriter.renderMarshallEventHeader(memberName: String, member: MemberShape, target: Shape) {
+        val headerName = member.memberName
+        handleOptional(
+            symbolProvider.toSymbol(member).isOptional(),
+            "inner.$memberName",
+            "value",
+            { input -> renderAddHeader(headerName, input, target) },
+        )
+    }
+
+    private fun RustWriter.renderAddHeader(headerName: String, inputName: String, target: Shape) {
+        withBlock("headers.push(", ");") {
+            rustTemplate("#{Header}::new", *codegenScope)
+            withBlock("(", ")") {
+                rust("${headerName.dq()},")
+                rustTemplate("#{HeaderValue}::", *codegenScope)
+                when (target) {
```

Comment:
maybe add a link to the list of valid header shapes: https://awslabs.github.io/smithy/1.0/spec/core/stream-traits.html#eventheader-trait

(since float, which I would expect to be supported, is not)